Add /cycle command (a.k.a. /rejoin)

dgw opened the pull request:
First pass at implementing #1436, no idea if I'm doing this right but gotta start somewhere! I expect this branch will be force-pushed approximately 5,821 times before being merged, because I still don't actually understand how commands work.

Tested locally under Windows inside Git Bash (which doesn't terminate `node.exe` on <kbd>Ctrl</kbd>+<kbd>C</kbd> :disappointed:). Line endings should be fine because Git's `core.autocrlf` setting is enabled, but the linter also reports unused arguments (not sure if I can define the new function with just `network` and `chan`).

For a first pass it works pretty great, just as expected except the part message isn't used (doesn't appear to be used by `/part` itself either; when I tested, the part message sent was `#channelname` for either command). I'd like to make it so `/cycle`ing a channel doesn't move it to the bottom of the list, but so many PRs end up with bikeshedding that I didn't want to start out by bikeshedding myself. :laughing:

astorije commented on src/plugins/inputs/cycle.js:
```diff
@@ -0,0 +1,31 @@
+"use strict";
```
Mind calling the file `rejoin.js` please? :) I'd like this to be the canonical name, and `/cycle` seen as an alias.

dgw replied:
I disagree, on the basis that `/cycle` is a long-standing name for the function (and is even a builtin on some networks, e.g. [SwiftIRC](https://wiki.swiftirc.net/wiki/Other_commands#Cycle.2C_Hop)). "rejoin" is usually attached to rejoining channels upon being kicked.

Input from other maintainers (@xPaw, @YaManicKill?) will be the deciding factor here, though. If there's consensus to canonically call this `rejoin`, I won't fight it. But it is my strongly held belief that calling it `rejoin` instead of `cycle` is more or less breaking from tradition for the sake of doing so.

astorije replied:
I understand where you're coming from, and I don't disagree with the rationale in terms of usage which is why I'm happy with a `/cycle` alias, but when it comes to the filenames and such, this is The Lounge codebase territory.
We don't have other command-related filenames with `de`/`re` suffixes yet, but we do have commands that have their counterparts prefixed with `de` (`voice` vs. `devoice`, etc.). This file should then be `rejoin.js`, with support for `cycle` command.

astorije commented:
If you look into `index.html`, you'll see a help section with commands, could you add your `/rejoin` command + `/cycle` alias there, please? You'll find example on other commands with aliases there.

dgw commented:
I will update the docs once the issue of canonical name has been settled. (And I'll also port my changes to a PR on the docs repo, once this is merged.)

astorije commented:
> except the part message isn't used (doesn't appear to be used by `/part` itself either; when I tested, the part message sent was `#channelname` for either command)

Interestingly, I'm encountering the same issue when testing #1448. Not sure what's going on, but I wouldn't worry about that for now.

> (And I'll also port my changes to a PR on the docs repo, once this is merged.)

If you're talking about [this page](https://thelounge.github.io/docs/client/commands.html), don't worry about it, it's deprecated and was removed from the new docs (WIP). Help window is the source of truth.

astorije commented on src/plugins/inputs/cycle.js:
```diff
@@ -0,0 +1,31 @@
+
+exports.commands = ["cycle", "rejoin"];
+
+exports.input = function(network, chan, cmd, args) {
```
Yep, you can just do:

```js
exports.input = function(network, chan) {
```

dgw commented:
Haven't squashed yet, but I added the help and renamed the command file.

Still want to figure out how to make it cycle without destroying the existing channel and losing scrollback.

xPaw commented on src/plugins/inputs/rejoin.js:
```diff
@@ -0,0 +1,31 @@
+	if (chan.type !== Chan.Type.CHANNEL) {
+		chan.pushMessage(this, new Msg({
+			type: Msg.Type.ERROR,
+			text: "You can only cycle channels."
```
s/cycle/rejoin

xPaw commented on src/plugins/inputs/rejoin.js:
```diff
@@ -0,0 +1,31 @@
+	});
+	network.irc.join(chan.name);
+
+	this.save();
```
Don't need this either.

dgw commented:
Squashed

astorije commented on src/plugins/inputs/rejoin.js:
```diff
@@ -0,0 +1,31 @@
+		return;
+	}
+
+	this.emit("part", {
```
We can't keep the chat window open without losing history like @dgw was trying to achieve, right?

xPaw replied:
Not sending emits to client *will* keep history.

astorije replied:
Odd, I tested this PR and it didn't.

astorije replied:
@dgw, could you confirm this? When I tested, it killed the window and history then recreated it.

dgw replied:
Sadly I cannot confirm that it keeps history. Even after making sure git said I'm up-to-date with `origin/cycle-command`, making sure the `emit`s were not present in `rejoin.js`, and running a fresh `npm run build`, starting up a test instance and checking the command still cleared the channel history.

I'm confused. Is the server perhaps catching the `PART` command and destroying the channel on its end, which then propagates to the client and clears the history?

astorije commented on client/index.html:
```diff
@@ -787,6 +787,19 @@ <h1 class="title">Help</h1>
+							<div class="description">
+								<p>
+									Equivalent to doing <code>/part</code> followed immediately
+									by <code>/join #channel</code>.
```
I'd prefer something like "Reload the current channel by leaving it then immediately re-joining it."

dgw replied:
I've been toying with wordings. Do any of these strike your fancy?

````
Quickly part and rejoin the current channel. Effectively
reloads the user list and syncs modes, at the expense of
losing channel scrollback.
----
Shortcut to leave and then immediately rejoin the current
channel, reloading users/modes/etc.
----
Leaves and immediately rejoins the current channel. Useful
to quickly get op from ChanServ in an empty channel, for example.
````

I'm not convinced that the verbiage about losing history is needed, but I think it's helpful unless a way to keep history can be implemented. :\

astorije replied:
Good stuff @dgw (sorry I missed that notification)! 👏

I like a mix of first and last:

```
Leave and immediately rejoin the current channel. Useful
to quickly get op from ChanServ in an empty channel, for example.
```

> I'm not convinced that the verbiage about losing history is needed

I agree with that part, let's not.

xPaw commented:
Damn yeah I forgot that receiving part from the server will close the channel.

https://github.com/thelounge/lounge/blob/6ba2b8a3d1440d46252e70e3fc11db622669eef8/src/plugins/irc-events/part.js#L14-L20

We might also hit an issue here if we ignore the part:
https://github.com/thelounge/lounge/blob/6ba2b8a3d1440d46252e70e3fc11db622669eef8/src/plugins/irc-events/join.js#L25

astorije commented:
Is there anything we can do about this?

YaManicKill commented:
> Is there anything we can do about this, you think?

Well yes, it depends how you want it to work. 

We could add a command (or an extra bit to the part command) that will part a channel, but not hide it or remove history. Could be useful.

dgw commented:
I noticed that receiving a kick doesn't destroy the channel. It does clear the list of users, but that's part of the point of this command (one use is to get ChanServ to sync up your modes). I might play with that, though it will probably lead to nothing.

dgw commented:
Updated help text and re-squashed.

Is this in merge-now-and-improve-later status, or in blocked-on-improvements-to-part-command status?

YaManicKill commented:
I think we should merge, as it just simplifies the `/part` and `/join #chan` process, and we can look into making it not lose scrollback later. It's still an improvement, it doesn't break any existing functionality.

YaManicKill commented:
@astorije Any objections to this being merged in the current state and adding an issue to look at keeping scrollback?

astorije commented:
> @astorije Any objections to this being merged in the current state and adding an issue to look at keeping scrollback?

Not at all, that was my feeling too :)

> If merged as-is, I would like to create the followup issue myself. Taking ownership of the "mess" I made, kinda. 😀

That's a great mindset, thank you!! 👏 👏
